### PR TITLE
Implement Grid

### PR DESCRIPTION
**Pull request by khaletska**
- Added usage of Grid for dynamically aligning cards on the screen
- Deal more cards button works until the end of the deck
- Made a HUGE mess 😢

### CODE REVIEW

**Comment by khaletska** [comment]
Found a bug: If the match id chosen and highlighted, next chosen card disappears too (if the deck is empty)

**Comment by Shiaulis** [comment]
> Found a bug: If the match id chosen and highlighted, next chosen card disappears too (if the deck is empty)

does it mean you are going to fix it?

**Comment by khaletska** [comment]
> > Found a bug: If the match id chosen and highlighted, next chosen card disappears too (if the deck is empty)
> 
> does it mean you are going to fix it?

Fixed

File: Set/Set/Controller/ViewController.swift [hunk 1]
```diff
@@ -29,33 +29,28 @@ final class ViewController: UIViewController {
         startNewGame()
     }
 
-    @IBAction private func cardButtonTapped(_ button: CardButton) {
-        if let index = self.cardButtons.firstIndex(of: button) {
-            self.game.touchCard(index: index)
-        }
+    private func cardButtonTapped(_ index: Int) {
+//        if (self.game.shownCards.compactMap { $0 }[index] != nil) {
```
Comment (Shiaulis) [inline Set/Set/Controller/ViewController.swift]: let's remove all commented code 🙏🏾

File: Set/Set/View/GridView.swift [hunk 2]
```diff
@@ -9,12 +9,57 @@ import UIKit
...
+        self.subviews.forEach { $0.removeFromSuperview() }
+        self.calculator.cellCount = cards.count
+        self.calculator.frame = self.bounds
+
+        for (index, card) in cards.enumerated() {
+            guard let frame = self.calculator[index] else {
+                continue
+            }
+
+            let cardButtonView = CardButton(frame: frame)
```
Comment (Shiaulis) [inline Set/Set/View/GridView.swift]: if cardButton is based on on card it makes sense to pass card within init and setup button based on the card itself

File: Set/Set/Model/SetGame.swift [hunk 3]
```diff
@@ -14,21 +14,26 @@ enum CardState {
...
-    private var shownCards: Array<Card?> = .init(repeating: nil, count: 24)
     private var chosenCardsIndices: Array<Int> = []
     private var matchedCardsIndices: Array<Int> = []
     private var unmatchedCardsIndices: Array<Int> = []
     private(set) var score: Int = 0
 
-    func touchCard(index: Int) {
-        guard !self.matchedCardsIndices.contains(index) else {
+    func touchCard(indexOnBoard: Int) {
+        guard let ID = getCardID(by: indexOnBoard) else {
```
Comment (Shiaulis) [inline Set/Set/Model/SetGame.swift]:
```suggestion
        guard let id = getCardID(by: indexOnBoard) else {
```

File: Set/Set/Model/SetGame.swift [hunk 4]
```diff
@@ -61,22 +65,18 @@ final class SetGame {
         self.delegate.gameDidChange()
     }
 
-    func getCard(for index: Int) -> Card? {
-        guard index < shownCards.count else {
-            return nil
+    func getCardState(for indexOnBoard: Int) -> CardState? {
+        guard let ID = getCardID(by: indexOnBoard) else {
```
Comment (Shiaulis) [inline Set/Set/Model/SetGame.swift]:
```suggestion
        guard let id = getCardID(by: indexOnBoard) else {
```

File: Set/Set/Model/SetGame.swift [hunk 5]
```diff
@@ -14,21 +14,26 @@ enum CardState {
...
-        !self.deck.isEmpty && (findEmptySlot() != nil || self.matchedCardsIndices.count == 3)
+        !self.deck.isEmpty
     }
     var delegate: SetGameDelegate!
+    var shownCards: Array<Card?> = .init(repeating: nil, count: 12)
     private var deck: Array<Card> = SetGame.generateDeck()
-    private var shownCards: Array<Card?> = .init(repeating: nil, count: 24)
     private var chosenCardsIndices: Array<Int> = []
     private var matchedCardsIndices: Array<Int> = []
     private var unmatchedCardsIndices: Array<Int> = []
```
Comment (Shiaulis) [inline Set/Set/Model/SetGame.swift]: i assume it is not indices but IDs?

File: Set/Set/Model/SetGame.swift [hunk 6]
```diff
@@ -86,17 +86,19 @@ final class SetGame {
...
-                        deck.append(Card(color: color, shade: shade, shape: shape, number: number))
+                        deck.append(Card(color: color, shade: shade, shape: shape, number: number, ID: iterator))
+                        iterator += 1
                     }
                 }
             }
         }
 
-        return deck.shuffled()
+        return deck//.shuffled()
```
Comment (Shiaulis) [inline Set/Set/Model/SetGame.swift]: i would prefer deck to be shuffled

**Review by Shiaulis — CHANGES_REQUESTED** [review]
left some comments.
Make sure to review your own code in Github before requesting review from others. It is the easiest way to spot such issues that unnecessary commented code or some typos

**Comment by khaletska** [comment]
> such issues that unnecessary commented code or some typos

sorry about this

File: Set/Set/Model/SetGame.swift (hunk 6 above)
Comment (khaletska) [reply Set/Set/Model/SetGame.swift]: ☢️

File: Set/Set/Controller/Card.swift [hunk 7]
```diff
@@ -12,6 +12,7 @@ struct Card {
     let shade: Shade
     let shape: Shape
     let number: Number
+    let id: Int
```
Comment (Shiaulis) [inline Set/Set/Controller/Card.swift:15]: In future I would recommend you to use `Identifiable` protocol and access object ID  type by `Card.ID` to abstract its implementation

**Review by Shiaulis — APPROVED** [review]
left comment regarding future usage of `Identifiable`.
Another comment (no need to fix it) is that we should not limit size of a button if it has potential to grow. With current constraints but with bigger font current layout becomes broken 

image

image

**Pull request merged**